projects/components/src/utils/test/widget-object/widget-object.ts

```diff
@@ -2,111 +2,220 @@
-    [P in keyof T]: T[P] extends (...args: any[]) => infer CR
-        ? (...args: Parameters<T[P]>) => ConvertGenericOnly<CR, R>
-        : ConvertGenericOnly<T[P], R>;
-};
+export class BaseWidgetObject<T> {
+    /**
+     * Helpers for create methods to find HTML elements within the widget object. They can find two types of elements:
+     *
+     * - WidgetObjectElements<T> which wrap the T which is unknown to this base class into a common interface that
+     *   can be be used within to WidgetObject subclasses to provide methods that can abstract interactions with
+     *   the widget. For example, a `loginWidget.login(usr, pwd)` call which types into the user and password fields
+     *   and then hits the submit button; These elements don't provide means to assert anything or even to read
+     *   information from the widget objects elements such as their text, CSS class names, etc.
+     *
+     * - The generic specified by <T> which depends on the implementation such as a DebugElement in Angular, a
+     *   Chainable in Cypress, or a WebElement in Selenium.
+     *
+     * Note that unlike Google Material's Harnesses, we do not impose a `TestElement` interface to be used from
+     * different implementations where all the methods return promises. The main reason is that it allows
+     * calls from different implementations to feel more natural, according to their testing environment. For example:
+     *
+     * - Avoids adding `await`s all over unit tests where they aren't necessarily needed. When trying this approach,
+     *   it was tricky to try to convert Cypress values into Promises in a reliable way. It initially looked promising
+     *   (pun intended) but turned out not to work with nested promises.
+     *
+     * - Allows the calls to WidgetObject methods to be more like their native objects instead of forcing them
+     *   to comply with a TestElement interface we have defined. That is, Cypress has a natural way to read text at
+     *   polling intervals automatically so you can still benefit from that when writing a Cypress Tests while
+     *   you would need a different mechanism in your Angular unit tests where you have a DebugElement.
+     *
+     *   This also prevents widget object methods from querying information about the elements forcing any assertions
+     *   to be run from the outside, which we believe is the correct way to structure codes since assertions are
+     *   framework specific.
+     *
+     *   Each factory method returns an ElementLocator<WidgetObjectElement<T>> which should never be public
+     *   but also has a corresponding method prefixed with _ to indicate the it returns `T` which is what should be
+     *   consumed by users of the class.
+     */
+    protected factory = {
+        /**
+         * Utility to create versions of factories that return the unwrapped T
+         */
+        unwrap: (fun: ElementLocator<WidgetObjectElement<T>>) => {
```

Comment:
I think this is useful to leave here but I'm not sure anyone is really going to use it. I think most people will end up creating the duplication. 

But a bigger question is how do we prevent people from exposing WidgetObjectElements from their widget objects?

Also, I'm not sure the names `_css` and `css` make it totally clear what's the difference between the two besides one's an internal call? Can you rename `_css` to `_cssWidgetInternal`? This makes it clear that it's returning a widget and extra clear that it should be only used internally?